### PR TITLE
CR-317 Reworked file structure 

### PR DESCRIPTION
**Pull request by RaygunDaniel**
We have separated the Crash reporting and RUM logic into separate classes with the RaygunClient class now acting as a wrapper through which the user can interact with both. We have also altered function names and variable names to be more robust as well as removing redundant variables and methods and concatenating files where possible (Transport.ts and Helper.ts have been dissolved).

We have not completed our rework of the Real User Monitoring code as we feel that this will be a large enough scope to warrant a separate PR.

### CODE REVIEW

File: sdk/src/RaygunClient.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+const updateCustomData = (updater: (customData: CustomData) => CustomData) => {
+  if (CrashReportingUnavailable()) return;
+  crashReporter.updateCustomData(updater);
+}
+
+/**
+ * Checks whether or not the user has initialised the client AND enabled crash reporting.
+ * @constructor
+ */
+const CrashReportingUnavailable = () => {
```
Comment (mduncan26) [inline sdk/src/RaygunClient.ts:184]: It is preferable to have this method return a boolean and leave the logging to the method that called it. The logging can then be specific to the feature the customer is attempting to use at the wrong time. This helps customer narrow down where they are doing things out of order. They could have a race condition where they are setting a breadcrumb for an initial view before initing the client and specific logs help with this.

File: sdk/src/RaygunClient.ts [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+  realUserMonitor.sendCustomRUMEvent(
+    Options.apiKey,
+    eventType,
+    name,
+    timeUsedInMs,
+    ''
+  );
+};
+
+const RealUserMonitoringUnavailable = () => {
```
Comment (mduncan26) [inline sdk/src/RaygunClient.ts:209]: Same as the CrashReportingUnavailable comment above

File: sdk/src/RealUserMonitor.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
+import {RealUserMonitoringEvents, Session, User} from "./Types";
+import {NativeEventEmitter, NativeModules, Platform} from 'react-native';
+import {setupNetworkMonitoring} from "./NetworkMonitor";
+import {getDeviceBasedId, log, warn} from "./Utils";
+
+const {RaygunNativeBridge} = NativeModules;
+const {osVersion, platform} = RaygunNativeBridge;
+
+const defaultURLIgnoreList = ['api.raygun.com', 'localhost:8081/symbolicate'];
+const SessionRotateThreshold = 30 * 60 * 100;
```
Comment (mduncan26) [inline sdk/src/RealUserMonitor.ts:10]: Best to put a comment on what this value evaluates to in time (seconds, minutes etc)

File: sdk/src/Types.ts [hunk 4]
```diff
@@ -104,51 +108,38 @@ export interface CrashReportPayload {
...
+// REAL USER MONITORING SPECIFIC TYPES
+//-------------------------------------------------------------------------------------------------
 
-enum RUMEventTypes {
+export enum RealUserMonitoringEvents {
   SessionStart = 'session_start',
   SessionEnd = 'session_end',
-  Timing = 'mobile_event_timing'
+  EventTiming = 'mobile_event_timing',
+  ActivityLoaded = 'p',
```
Comment (mduncan26) [inline sdk/src/Types.ts:119]: ActivityLoaded should be renamed to ViewLoaded as Activity is only relevant to Android.

**Review by mduncan26 — CHANGES_REQUESTED** [review]
Some things to go through and update. There are other things in there we will need to change but we can go through that at a later date.

File: README.md [hunk 5]
```diff
@@ -69,8 +69,8 @@ RaygunClient.init({
   //RaygunClientOptions
   apiKey: YOUR_RAYGUN_APIKEY,
   version: YOUR_APP_VERSION,
-  enableNativeCrashReporting: true,
-  enableRUM: true,
+  disableNativeCrashReporting: false,
```
[the remark this answers is not in the document]
Comment (RaygunDaniel) [reply README.md:72]: Addressed

File: sdk/src/CrashReporter.ts [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+  private RAYGUN_CRASH_REPORT_ENDPOINT = 'https://api.raygun.com/entries';
+
+  constructor(curSession: Session,
+              apiKey: string,
+              disableNativeCrashReporting: boolean,
+              customCrashReportingEndpoint: string,
+              onBeforeSendingCrashReport: BeforeSendHandler | null,
+              version: string) {
+
+    //Setup error handler to divert errors to crash reporter
```
[the remark this answers is not in the document]
Comment (RaygunDaniel) [reply sdk/src/CrashReporter.ts:44]: Donezo

File: sdk/src/CrashReporter.ts [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+      allRejections: true,
+      onUnhandled: this.processUnhandledRejection
+    });
+
+
+    if (disableNativeCrashReporting) {
+      setTimeout(() => this.sendCachedReports(apiKey, customCrashReportingEndpoint), 10);
+    }
+
+    // Assign the values parsed in (assuming initiation is the only time these are altered).
```
[the remark this answers is not in the document]
Comment (RaygunDaniel) [reply sdk/src/CrashReporter.ts:65]: Donezo

File: sdk/src/RaygunClient.ts [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+ * Create a breadcrumb in the current session.
+ * @param message
+ * @param details
+ */
+const recordBreadcrumb = (message: string, details?: BreadcrumbOption) => {
+  if (CrashReportingUnavailable()) return;
+  crashReporter.recordBreadcrumb(message, details);
+
+};
+const sendCustomError = async (error: Error, ...params: any) => {
```
[the remark this answers is not in the document]
Comment (RaygunDaniel) [reply sdk/src/RaygunClient.ts:156]: DONE

File: sdk/src/RaygunClient.ts (hunk 1 above)
Comment (RaygunDaniel) [reply sdk/src/RaygunClient.ts:184]: Boosh

File: sdk/src/CrashReporter.ts [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+    this.apiKey = apiKey;
+    this.disableNativeCrashReporting = disableNativeCrashReporting;
+    this.customCrashReportingEndpoint = customCrashReportingEndpoint;
+    this.onBeforeSendingCrashReport = onBeforeSendingCrashReport;
+    this.version = version;
+
+  }
+
+//-------------------------------------------------------------------------------------------------
+// ALTERING SESSION DATA
```
[the remark this answers is not in the document]
Comment (RaygunDaniel) [reply sdk/src/CrashReporter.ts:76]: DONE

**Review by mduncan26 — APPROVED** [review]

**Pull request merged**